### PR TITLE
feat(CL): core swaprouter logic and tests (merge to main 3)

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

We would like to drop `concentrated-liquidity-main` branch to avoid branch synching overhead. As part of that, we are going to be incrementally merging the current concentrated liquidity feature state to `main`.

All this logic has already been given a round of reviews since we treated `concentrated-liqudiity-main` branch as `main` with appropriate review processes.

This particular PR does the following:
- fully merges `x/swaprouter` and its tests
- disables the failing tests that depend on gamm refactor
- inclides minor non-breaking API changes in GAMM

This is the final swaprouter PR. As soon as we merge gamm refactor, the swap router can be fully enabled.

For the latest state of the feature that we are merging, see: https://github.com/osmosis-labs/osmosis/tree/concentrated-liquidity-main

See spec: https://github.com/osmosis-labs/osmosis/pull/3052/files

### CODE REVIEW

File: osmoutils/store_helper_test.go [hunk 1]
```diff
@@ -811,6 +811,98 @@ func (s *TestSuite) TestMustGet() {
 	}
 }
 
+// TestMustGet tests that GetIfFound returns a boolean indicating
```
Comment (czarcas7ic) [inline osmoutils/store_helper_test.go:814]:
```suggestion
// TestGetIfFound tests that GetIfFound returns a boolean indicating 
```

File: x/swaprouter/client/cli/cli_test.go [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+	testCases := []struct {
+		name string
+		args []string
+
+		expectErr    bool
+		respType     proto.Message
+		expectedCode uint32
+	}{
+		{
+			"swap exact amount out", // osmosisd tx swaprouter swap-exact-amount-out 10stake 20 --swap-route-pool-ids=1 --swap-route-denoms=node0token --from=validator --keyring-backend=test --chain-id=testing --yes
```
Comment (czarcas7ic) [inline x/swaprouter/client/cli/cli_test.go:95]: do we intend on adding test cases for multi route CLI swaps?

File: x/swaprouter/client/cli/query.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+// GetCmdEstimateSwapExactAmountIn returns estimation of output coin when amount of x token input.
+func GetCmdEstimateSwapExactAmountIn() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "estimate-swap-exact-amount-in <poolID> <sender> <tokenIn>",
+		Short: "Query estimate-swap-exact-amount-in",
+		Long: strings.TrimSpace(
+			fmt.Sprintf(`Query estimate-swap-exact-amount-in.
+Example:
+$ %s query swaprouter estimate-swap-exact-amount-in 1 osm11vmx8jtggpd9u7qr0t8vxclycz85u925sazglr7 stake --swap-route-pool-ids=2 --swap-route-pool-ids=3
+`,
```
Comment (czarcas7ic) [inline x/swaprouter/client/cli/query.go:45]: would like to test this over multiple routes

File: x/swaprouter/client/cli/query.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+// GetCmdEstimateSwapExactAmountOut returns estimation of input coin to get exact amount of x token output.
+func GetCmdEstimateSwapExactAmountOut() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "estimate-swap-exact-amount-out <poolID> <sender> <tokenOut>",
+		Short: "Query estimate-swap-exact-amount-out",
+		Long: strings.TrimSpace(
+			fmt.Sprintf(`Query estimate-swap-exact-amount-out.
+Example:
+$ %s query swaprouter estimate-swap-exact-amount-out 1 osm11vmx8jtggpd9u7qr0t8vxclycz85u925sazglr7 stake --swap-route-pool-ids=2 --swap-route-pool-ids=3
+`,
```
Comment (czarcas7ic) [inline x/swaprouter/client/cli/query.go:98]: same

File: x/swaprouter/msg_server.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+func (server msgServer) CreateStableswapPool(goCtx context.Context, msg *stableswap.MsgCreateStableswapPool) (*stableswap.MsgCreateStableswapPoolResponse, error) {
+	poolId, err := server.CreatePool(goCtx, msg)
+	if err != nil {
+		return nil, err
+	}
+	return &stableswap.MsgCreateStableswapPoolResponse{PoolID: poolId}, nil
+}
+
+// func (server msgServer) StableSwapAdjustScalingFactors(goCtx context.Context, msg *stableswap.MsgStableSwapAdjustScalingFactors) (*stableswap.MsgStableSwapAdjustScalingFactorsResponse, error) {
```
Comment (czarcas7ic) [inline x/swaprouter/msg_server.go:61]: TODO: re-enable after gamm refactor is ported

File: x/swaprouter/types/routes.go [hunk 6]
```diff
@@ -9,22 +9,22 @@ import (
...
 	NewAccount(sdk.Context, authtypes.AccountI) authtypes.AccountI
 	GetAccount(ctx sdk.Context, addr sdk.AccAddress) authtypes.AccountI
 	SetAccount(ctx sdk.Context, acc authtypes.AccountI)
 }
 
-// BankI defines the banking contract that must be fulfilled when
-// creating a x/swaprouter keeper.
+// BankKeeper defines the banking contract that must be fulfilled when
+// creating a x/gamm keeper.
 type BankI interface {
```
Comment (czarcas7ic) [inline x/swaprouter/types/routes.go:22]: These should say AccountI and BankI as well as x/swaprouter I believe

File: x/swaprouter/types/routes.go [hunk 7]
```diff
@@ -9,22 +9,22 @@ import (
...
 	SendCoins(ctx sdk.Context, fromAddr sdk.AccAddress, toAddr sdk.AccAddress, amt sdk.Coins) error
 	SetDenomMetaData(ctx sdk.Context, denomMetaData banktypes.Metadata)
 }
 
-// CommunityPoolI defines the contract needed to be fulfilled for distribution keeper.
+// CommunityPoolKeeper defines the contract needed to be fulfilled for distribution keeper.
```
Comment (czarcas7ic) [inline x/swaprouter/types/routes.go:27]:
```suggestion
// CommunityPoolI defines the contract needed to be fulfilled for distribution keeper.
```

**Review by czarcas7ic — APPROVED** [review]
👍

File: x/swaprouter/client/cli/cli_test.go (hunk 2 above)
Comment (p0mvn) [reply x/swaprouter/client/cli/cli_test.go:95]: These have been ported from gamm a couple of weeks ago. Have there been any new test cases that we added that were missed here?

File: x/swaprouter/msg_server.go (hunk 5 above)
Comment (p0mvn) [reply x/swaprouter/msg_server.go:61]: I think this actually needs to be removed, thanks for catching

File: x/swaprouter/types/routes.go (hunk 6 above)
Comment (p0mvn) [reply x/swaprouter/types/routes.go:22]: Ugh, I keep merging directly merging from `concentrated-liqudity-main` that has an incorrect state. Sorry for the same issues again

**Pull request merged**